### PR TITLE
Add JSON instances

### PR DESCRIPTION
**Pull request by willbasky**

### CODE REVIEW

**Comment by Mikolaj** [comment]
Thank you for the PR. I'm torn on this one. It's certainly useful and avoids orphan instances, but OTOH we mimic package containers and it doesn't have package aeson dependency (which depends on a lot of other packages). OTOH, we can't expect aeson to include instances for these types until they are finally included, in one version or another, in package containers, so perhaps including them here is a convenient temporary measure. What are your thoughts?

**Comment by willbasky** [comment]
```
we can't expect aeson to include instances for these types until they are finally included, in one version or another, in package containers, 
```
1. included to `enummapset` or to `containers`?
2. What policy or condition `aeson` follows to add new instances? Do they see instances here and move them unto oneself? Or...

in anyway, I was afraid `aeson` dependency extends dependency list and drops simplicity, it was my reason why JSON not included. Actual reason to be like  `containers`. 
Well, `containers` has instances in `aeson`, unlike `enummapset`. Hence, to be similar `enummapset` has to have instances somehow. Like #17 or via contribute to `aeson`. The last case demand an answer on 2 question.

**Comment by Mikolaj** [comment]
1. I hope, at some point,  `enummapset` types (or their variant from another package) are included in `containers`. Then, I suppose, `aeson` will include their instances.
2. That's a good question. Let's ask the kind `aeson` maintainers. Note that at this time, they would need to chose between `enummapset`, `enummaps` and `enummapset-th` (at least). Myself, I can't tell which of these packages is better. None of them currently includes the instances (unless they are in separate packages, but I doubt that). Perhaps they just accept whichever package submits a PR and then change the instances once one of the packages gets standardized into base/containers/whatever?

@bergmark, @phadej, what are your thoughts?

File: Data/EnumSet.hs
Comment (phadej) [inline Data/EnumSet.hs:126]: Does this use `ToJSON a`, or just encodes a list of integers?

Write tests, not only roundtrip properties, but also examples of what the encodings look like.

File: Data/EnumMap/Base.hs
Comment (phadej) [inline Data/EnumMap/Base.hs:206]: `FromJSON1` and `FromJSON2` instances may be useful too.

**Review by phadej — COMMENTED** [review]
Re: adding dependencies to `aeson`.

We have been conservative when adding dependencies, but we added `these` and `strict`. Hard to say without looking into `enummapset` more closely.

File: Data/EnumSet.hs
Comment (willbasky) [reply Data/EnumSet.hs:126]: Yes, it encode a list of integers at the end. 
I ll add tests.

File: Data/EnumMap/Base.hs
Comment (willbasky) [reply Data/EnumMap/Base.hs:206]: I ll check it

**Comment by willbasky** [comment]
@Mikolaj My project `enummaps`, i was going to update stack's lts and found it has no repo on github. I found your library approach very similar and it presented on github as well. I not considered `enummapset-th` because of `th` :)

**Comment by phadej** [comment]
The fact there are three packages doing the same thing (and none being de facto package for the job) makes it more difficult for `aeson` to depend on one of them.

**Comment by willbasky** [comment]
@phadej I guess, it is not first time you faced with several similar packages. What options do you have to resolve it?
1. Three maintainers merged projects to one.
2. Vote for the best
3.  ...?

**Comment by phadej** [comment]
@willbasky depends

One example is`data-fix`: it already existed but wasn't "best", `recursion-schemes` had own `Fix`. We merged functionality, making `recursion-schemes` depend on `data-fix`. That process is still "ongoing" as there are `Fix`es around on Hackage still.

`these` and `data-or` is probably an outlier, as `data-or` weren't really used by anyone, and `these` is by `reflex`, so there was "de facto" package, but it took some time (years) to arrive to its current state.

The common about these is that such "consolidation" doesn't happen over a week or months. Making users of `enummapsset-th` use `enummapset` could be good start. Figuring out if `enummaps` has something `enummapset` doesn't, and adding to latter. And overall improvement of `enummapset` to be top-tier package (starting with having working CI) would help with lurking more users. (`enummapset` is good candidate, as `enummapset-th` needs TH, which is not good, and `enummaps` doesn't seem to have users on Hackage).

**Comment by Mikolaj** [comment]
> And overall improvement of `enummapset` to be top-tier package (starting with having working CI) would help with lurking more users.

How dare you. Oh, travis is now kaput even for small projects. :( PRs welcome. :)

**Comment by Mikolaj** [comment]
@willbasky: thank you for the tests. Do you think you could try if the automatic way of adding GHA CI via https://github.com/haskell-CI/haskell-ci works here so that your tests are exercised properly at each push? I'm ready to assist and also to give you whatever project permissions are needed.

**Comment by willbasky** [comment]
I could try `[haskell-ci](https://github.com/haskell-CI/haskell-ci)`

btw, JSON1 and JSON2 left to add, never did it before.

**Comment by willbasky** [comment]
@Mikolaj  I think you want to add a `haskell-ci` script extension to Action.

**Comment by Mikolaj** [comment]
Doh, I can't find where to turn it on. I've invited you to co-maintain the repo (I hope these are the permissions that the invite gives),

**Comment by Mikolaj** [comment]
BTW, the list of tested-with: GHC's is probably quite outdated. Feel free to remove old ones, add new and, generally, only consider last point version of each main release line. Then regenerate haskell-ci.

**Comment by willbasky** [comment]
> Feel free to remove old ones
oh, it eases a way to resolve it :)

**Comment by Mikolaj** [comment]
Actually, I think the travis script can be safely removed. Ee now have GHA and also travis no longer works (easily, at least) and I had quite bad interaction with Travis support after their change of policies.

**Comment by willbasky** [comment]
We could add JSON1 and 2 later. 
Test [passed](https://github.com/willbasky/enummapset/actions/runs/23238638310
I assume, actions not shown currently here because of I push commit through my repo.

**Comment by Mikolaj** [comment]
Yes, I think CI will be set up only once the PR is merged. Some setting flip may or may not then be needed.

I think once you are ready, we can just merge, and postpone the deccsion whether to keep the aeson instances in the version uploaded to Hackage. Perhaps before we release, the instances get miraculously merged to aeson. Or perhaps users would object to the new deps and explain how that hurts their use case. Or perhaps nobody objects and that tells use something, as well.

**Comment by willbasky** [comment]
> the instances get miraculously merged to aeson


For that some requirements are still wanted as @phadej asked (json1-2 and get the repo be ripened.
Let see.

**Review by Mikolaj — APPROVED** [review]

**Comment by Mikolaj** [comment]
Right. We have time. There is no major problem that forces us to release to Hackage ASAP. Unless you need the instances on Hackage ASAP.

Please feel free to merge and address the remaining issues in another PR.

**Comment by Mikolaj** [comment]
@willbasky: thank you!

**Comment by Mikolaj** [comment]
Lots of time passed and nothing emerged that would make the decision easier, so I just went ahead and published to Hackage. I waited until aeson supports text2 and tested locally in that configuration (but I haven't tested the instances, only compiling and running other functionality). Let's see what happens next.

**Pull request merged**
